### ISSUE TITLE
INTERNAL ASSERT FAILED at "../torch/csrc/autograd/python_torch_functions_manual.cpp":661 when returning a constant tensor in the forward method

### ISSUE THREAD

**Issue by alexandre-eymael** [pytorch/pytorch#132134]
### 🐛 Describe the bug

I am not sure of the exact cause of this bug, but when the output of the model involves a constant tensor (cf. example below), a RuntimeError is raised when trying to export it with TorchDynamo. I report this bug because the error message indicates that it is what I should do. If this behaviour is expected, I feel like the error message should be more descriptive.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F


class MyModel(nn.Module):

    def __init__(self):
        super(MyModel, self).__init__()
        self.conv1 = nn.Conv2d(1, 6, 5)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = F.max_pool2d(F.relu(self.conv1(x)), (2, 2))
        x = F.max_pool2d(F.relu(self.conv2(x)), 2)
        x = torch.flatten(x, 1)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        x = torch.tensor(1) # <-- Adding this line raises a 'RuntimeError: at::functionalization::impl::isFunctionalTensor(self_) INTERNAL ASSERT FAILED'
        return x
    
export_options = torch.onnx.ExportOptions(dynamic_shapes=True) # Even with `dynamic_shapes=False`, the error persists
torch_model = MyModel()
torch_input = torch.randn(1, 1, 32, 32)
onnx_program = torch.onnx.dynamo_export(torch_model, torch_input, export_options=export_options)
```

### Error logs

```
Traceback (most recent call last):
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/exporter.py", line 1428, in dynamo_export
    ).export()
      ^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/exporter.py", line 1171, in export
    graph_module = self.options.fx_tracer.generate_fx(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/fx/dynamo_graph_extractor.py", line 232, in generate_fx
    return self.pre_export_passes(options, model, graph_module, updated_model_args)  # type: ignore[return-value]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/fx/dynamo_graph_extractor.py", line 242, in pre_export_passes
    return exporter.common_pre_export_passes(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/exporter.py", line 1467, in common_pre_export_passes
    module = passes.Functionalize(
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/diagnostics/infra/decorator.py", line 151, in wrapper
    ctx.log_and_raise_if_error(diag)
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/diagnostics/infra/context.py", line 366, in log_and_raise_if_error
    raise diagnostic.source_exception
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/diagnostics/infra/decorator.py", line 135, in wrapper
    return_values = fn(*args, **kwargs)
                    ^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/fx/_pass.py", line 275, in run
    module = self._run(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/fx/passes/functionalization.py", line 123, in _run
    graph_module = proxy_tensor.make_fx(
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 1081, in wrapped
    t = dispatch_trace(wrap_key(func, args, fx_tracer, pre_dispatch), tracer=fx_tracer, concrete_args=tuple(phs))
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/_compile.py", line 24, in inner
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 451, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/_dynamo/external_utils.py", line 36, in inner
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 541, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 451, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/_dynamo/external_utils.py", line 36, in inner
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/fx/_symbolic_trace.py", line 793, in trace
    (self.create_arg(fn(*args)),),
                     ^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 559, in wrapped
    out = f(*tensors)
          ^^^^^^^^^^^
  File "<string>", line 1, in <lambda>
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/fx/passes/functionalization.py", line 95, in wrapped
    pytree.tree_map(torch._sync, out)
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/utils/_pytree.py", line 900, in tree_map
    return treespec.unflatten(map(func, *flat_args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/utils/_pytree.py", line 736, in unflatten
    leaves = list(leaves)
             ^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/_utils.py", line 887, in _functionalize_sync
    torch._functionalize_sync(t)  # type: ignore[attr-defined]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: at::functionalization::impl::isFunctionalTensor(self_) INTERNAL ASSERT FAILED at "../torch/csrc/autograd/python_torch_functions_manual.cpp":661, please report a bug to PyTorch. 

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/a/cedie/surya_onnx/playgrounds/onnx_test.py", line 31, in <module>
    onnx_program = torch.onnx.dynamo_export(torch_model, torch_input, export_options=export_options)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/a/.miniforge3/envs/surya_onnx/lib/python3.11/site-packages/torch/onnx/_internal/exporter.py", line 1439, in dynamo_export
    raise OnnxExporterError(
torch.onnx.OnnxExporterError: Failed to export the model to ONNX. Generating SARIF report at 'report_dynamo_export.sarif'. SARIF is a standard format for the output of static analysis tools. SARIF logs can be loaded in VS Code SARIF viewer extension, or SARIF web viewer (https://microsoft.github.io/sarif-web-component/). Please report a bug on PyTorch Github: https://github.com/pytorch/pytorch/issues
```

### Minified repro

```python
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
import torch.fx.experimental._config
torch._dynamo.config.specialize_int = True
torch._dynamo.config.assume_static_by_default = False
torch._dynamo.config.automatic_dynamic_shapes = False
torch._dynamo.config.capture_scalar_outputs = True
torch._dynamo.config.capture_dynamic_output_shape_ops = True

from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.L__self___conv1 = Conv2d(1, 6, kernel_size=(5, 5), stride=(1, 1))
        self.L__self___conv2 = Conv2d(6, 16, kernel_size=(5, 5), stride=(1, 1))
        self.L__self___fc1 = Linear(in_features=400, out_features=120, bias=True)
        self.L__self___fc2 = Linear(in_features=120, out_features=84, bias=True)
        self.L__self___fc3 = Linear(in_features=84, out_features=10, bias=True)

    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        l__self___conv1 = self.L__self___conv1(l_x_);  l_x_ = None
        relu = torch.nn.functional.relu(l__self___conv1);  l__self___conv1 = None
        x = torch.nn.functional.max_pool2d(relu, (2, 2));  relu = None
        l__self___conv2 = self.L__self___conv2(x);  x = None
        relu_1 = torch.nn.functional.relu(l__self___conv2);  l__self___conv2 = None
        x_1 = torch.nn.functional.max_pool2d(relu_1, 2);  relu_1 = None
        x_2 = torch.flatten(x_1, 1);  x_1 = None
        l__self___fc1 = self.L__self___fc1(x_2);  x_2 = None
        x_3 = torch.nn.functional.relu(l__self___fc1);  l__self___fc1 = None
        l__self___fc2 = self.L__self___fc2(x_3);  x_3 = None
        x_4 = torch.nn.functional.relu(l__self___fc2);  l__self___fc2 = None
        x_5 = self.L__self___fc3(x_4);  x_4 = None
        x_6 = torch.tensor(1)
        return (x_6,)

mod = Repro()

def load_args(reader):
    buf0 = reader.storage('3e9fbf7f2175dbf815b03cdb9328e36eaa8f0103', 4096)
    reader.tensor(buf0, (1, 1, 32, 32), is_leaf=True)  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='/home/a/cedie/surya_onnx/torch_compile_debug/run_2024_07_30_13_57_18_324762-pid_92116/minifier/checkpoints', autocast=False, backend=None)
```

### Versions

PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04 LTS (x86_64)
GCC version: (Ubuntu 13.2.0-23ubuntu4) 13.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.153.1-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070 Ti
Nvidia driver version: 560.70
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             24
On-line CPU(s) list:                0-23
Vendor ID:                          GenuineIntel
Model name:                         13th Gen Intel(R) Core(TM) i7-13700KF
CPU family:                         6
Model:                              183
Thread(s) per core:                 2
Core(s) per socket:                 12
Socket(s):                          1
Stepping:                           1
BogoMIPS:                           6835.20
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq vmx ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize flush_l1d arch_capabilities
Virtualization:                     VT-x
Hypervisor vendor:                  Microsoft
Virtualization type:                full
L1d cache:                          576 KiB (12 instances)
L1i cache:                          384 KiB (12 instances)
L2 cache:                           24 MiB (12 instances)
L3 cache:                           30 MiB (1 instance)
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.16.1
[pip3] onnxruntime==1.18.1
[pip3] onnxscript==0.1.0.dev20240730
[pip3] torch==2.3.0
[pip3] triton==2.3.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] triton                    2.3.0                    pypi_0    pypi

cc @ezyang @chauhang @penguinwu

**Comment by kit1980** [comment]
@justinchuby You added 2.5.0 milestone, do you have a fix for it?

**Comment by justinchuby** [comment]
The following with `dynamo=True` is successful with the latest torch nightly:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F


class MyModel(nn.Module):

    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(1, 6, 5)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = F.max_pool2d(F.relu(self.conv1(x)), (2, 2))
        x = F.max_pool2d(F.relu(self.conv2(x)), 2)
        x = torch.flatten(x, 1)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        x = torch.tensor(1)
        return x

torch_model = MyModel()
torch_input = torch.randn(1, 1, 32, 32)
onnx_program = torch.onnx.export(torch_model, (torch_input,), dynamo=True)
print(onnx_program)
```

You may leverage the new dynamic_shapes option to specify dynamic shapes too. The `dynamo_export` api currently runs on old logic, and will be updated to match `dynamo=True` in 2.6. Hope that helps.

**Comment by kit1980** [comment]
@justinchuby was it fixed before 2.5 branch cut? Any cherry-picks needed?

**Comment by justinchuby** [comment]
It should be fixed before the branch cut. I think we should be ok. Thanks for checking!
